File: src/test/kotlin/exercises/day2/java/Overloading.java
package exercises.day2.java;

public class Overloading {
    Point point1 = new Point();
    Point point2 = new Point();

    public void testThis(){
        point1.sum(point2);
    }

}


class Point {
    Integer a;
    Integer b;

    public Integer getA() {
        return a;
    }

    public void setA(Integer a) {
        this.a = a;
    }

    public Integer getB() {
        return b;
    }

    public void setB(Integer b) {
        this.b = b;
    }

    public Point sum(Point p) {
        Point result = new Point();
        result.setA(this.getA() + p.getA());
        result.setB(this.getB() + p.getB());
        return result;
    }

    public Point minus(Point p) {
        Point result = new Point();
        result.setA(this.getA() - p.getA());
        result.setB(this.getB() - p.getB());
        return result;
    }

    public Point times(Point p) {
        Point result = new Point();
        result.setA(this.getA() * p.getA());
        result.setB(this.getB() * p.getB());
        return result;
    }

    public Point div(Point p) {
        Point result = new Point();
        result.setA(this.getA() / p.getA());
        result.setB(this.getB() / p.getB());
        return result;
    }

    public Point inc() {
        Point result = new Point();
        result.setA(this.getA() + 1);
        result.setB(this.getB() + 1);
        return result;
    }

    public Point dec() {
        Point result = new Point();
        result.setA(this.getA() - 1);
        result.setB(this.getB() - 1);
        return result;
    }

}

File: src/test/kotlin/exercises/day2/kotlin/Overloading.kt
package exercises.day2.kotlin

import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.shouldBe

class Overloading : ShouldSpec({

    should("a list") {
        val point = Point(6, 8)
        val point2 = Point(3, 4)
        point + point2 shouldBe Point(9, 12)
        point - point2 shouldBe Point(3, 4)
        point * point2 shouldBe Point(18, 32)
        point / point2 shouldBe Point(2, 2)
        point.inc() shouldBe Point(7, 9)
        point.dec() shouldBe Point(5, 7)
    }

})

data class Point(val x: Int, val y: Int) {
    operator fun plus(p: Point) = Point(x + p.x, y + p.y)
    operator fun minus(p: Point) = Point(x - p.x, y - p.y)
    operator fun times(p: Point) = Point(x * p.x, y * p.y)
    operator fun div(p: Point) = Point(x / p.x, y / p.y)
    operator fun inc() = Point(x + 1, y + 1)
    operator fun dec() = Point(x - 1, y - 1)
}
